Noncopyable means no conformance to Copyable

kavon opened the pull request:
More work that is a NFC unless using `-enable-experimental-feature NoncopyableGenerics`

This change-set primarily overhauls the `isNoncopyable` methods of `TypeBase` and `ValueDecl`:

First, I've narrowed the scope of `ValueDecl::isNoncopyable` to `TypeDecl::isNoncopyable`. We are still defining `TypeDecl::isNoncopyable` as meaning "is always noncopyable" via checking for a `~Copyable` written on the decl. 
I'd eventually like this method to go away and/or be substantially changed into
1. `decl->isNoncopyable()` is rewritten as `getInterfaceType()->isNoncopyable()`
2. `TypeDecl::implicitCopyableBehavior()` returning one of `{None, Conditional, Always}` to reflect the different meanings of physically annotating the type with `~Copyable`.

For `TypeBase::isNoncopyable`, we now query whether the type conforms to `Copyable`. This change happens in coordination with actually synthesizing conformances to `Copyable` for all nominal types when it makes sense to (i.e., it doesn't have `~Copyable` written on it, hence why the legacy-form of `TypeDecl::isNoncopyable` lingers).

kavon committed:
[nfc] refactor `ValueDecl::isMoveOnly`

I've renamed the method to `TypeDecl::isNoncopyable`, because the query
doesn't make sense for many other kinds of `ValueDecl`'s beyond the
`TypeDecl`'s. In fact, it looks like no one was relying on that anyway.

Thus, we now have a distinction where in Sema, you ask whether
a `Type` or `TypeDecl` is "Noncopyable". But within SIL, we still
preserve the notion of "move-only" since there is additionally the
move-only type wrapper for types that otherwise support copying.

kavon committed:
[Sema] initial overhaul of `isNoncopyable`

This implementation has the function execute a request to scan the
inheritance clause of non-protocol nominals for a `~Copyable`. For
protocols, we look in the requirement signature.

This isn't our final state, as the GenericEnvironment needs to be
queried in general to determine of a Type is noncopyable. So for now
checking for a `~Copyable` only makes sense for Decls.

kavon committed:
[nfc] remove unused diagnostics

kavon committed:
[Sema] factor out the StorageVisitor pattern

This visitor is generally useful for other kinds of marker protocols
like `Copyable` that require structural checking over the storage of the
struct or enum.

kavon commented:
@swift-ci please smoke test

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -912,18 +913,81 @@ IsFinalRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
+
+  // For protocols, inspect its requirement signature for a Copyable requirement
+  if (auto *proto = dyn_cast<ProtocolDecl>(decl)) {
+    auto reqs = proto->getRequirementSignature().getRequirements();
```
This won't work if `Self: Copyable` is implied by some other `Self: P`. You can instead do this

```
protoDecl->getGenericSignature()->isRequiredProtocol(protoDecl->getSelfInterfaceType(), copyableProto)
```

kavon replied:
Fixed!

slavapestov commented on lib/Sema/TypeCheckInvertible.cpp:
```diff
@@ -0,0 +1,149 @@
+  if (isa<BuiltinTupleDecl>(nom))
+    llvm_unreachable("unexpected BuiltinTupleDecl");
+
+  if (auto strct = dyn_cast<StructDecl>(nom)) {
```
structDecl

slavapestov commented on lib/Sema/TypeCheckInvertible.cpp:
```diff
@@ -0,0 +1,149 @@
+    if (evaluateOrDefault(evaluator, IsNoncopyableRequest{nominal}, false))
+      return nullptr; // it's not Copyable.
+
+    // Otherwise, verify Copyable conformance.
```
You should just trust the type's inheritance clause here. Check that the stored properties are copyable in TypeCheckDeclPrimary.cpp's DeclChecker visitor

kavon replied:
I think you're right here in that I should just return the ProtocolConformance. But the checking of that conformance should probably happen in `TypeChecker::checkConformancesInContext` instead of at the top level / MiscDiagnostics.

kavon committed:
[nfc] factor out logic to remove inherited entry

kavon committed:
[Sema] "Noncopyable" means no Copyable conformance

kavon commented:
@swift-ci please smoke test macOS platform

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test

kavon committed:
[Sema] fix tuples and NoncopyableGenerics

The generic signature for the built-in TupleDecl was incorrect, stating
that the Element type must be Copyable. That in combination with an
extension that says tuples are Copyable when Element is copyable was
giving unconditional Copyable conformances for tuples that never
actually get checked.

Given that the existence of a conformance and no out-standing
conditional requirements is now what is used to determine if a value is
noncopyable, this was causing tuples with noncopyable elements to report
 that they are copyable in `isNoncopyable` requests.

kavon committed:
[Sema] diagnostic for conditional Copyable

We were allowing the catch-all "X is not compatible with generics yet"
diagnostic to be emitted when there's a failure to satisfy a requirement
for a conditional Copyable conformance. Instead, allow the existing
infrastructure to diagnose such failures-to-conform to emit a diagnostic
.

This patch also adds a check to catch future catch-all diagnostics in
asserts builds, since I'd like to phase out that diagnostic when
NoncopyableGenerics is enabled.

kavon committed:
[NoncopyableGenerics] misc fixes and improvements

kavon committed:
[NoncopyableGenerics] more regression testing

kavon committed:
[HasNoncopyableAnnotationRequest] handle `where`

Adds rudiementary support for searching `where` clauses for `~Copyable`
constraints, but the implementation is known to have flaws.

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7442,6 +7446,28 @@ ERROR(accessor_macro_not_single_var, none,
+      ())
+ERROR(noncopyable_type_member_in_copyable,none,
+      "%select{stored property %2|associated value %2}1 of "
+      "'Copyable'-conforming %kind3 has noncopyable type %0",
```
"has noncopyable stored property" maybe?

slavapestov commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -427,6 +428,25 @@ class IsMoveOnlyRequest
 };
 
+/// Determine whether the given type is noncopyable. Assumes type parameters
+/// have become archetypes.
```
"assumes a generic signature has been built" maybe?

slavapestov commented on lib/AST/ConformanceLookupTable.cpp:
```diff
@@ -810,7 +810,7 @@ bool ConformanceLookupTable::resolveConformances(ProtocolDecl *protocol) {
 }
 
-DeclContext *ConformanceLookupTable::getConformingContext(
+DeclContext *ConformanceLookupTable:: getConformingContext(
```
Stray space

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -157,22 +157,73 @@ bool TypeBase::isMarkerExistential() {
+  if (!hasArchetype() || hasOpenedExistential())
+  // TODO: the need for the following test is suspicious.
+    if (dc->getGenericEnvironmentOfContext() || !hasTypeParameter())
+      return dc->mapTypeIntoContext(this)->isNoncopyable();
```
mapTypeIntoContext() is a no-op if the type does not contain type parameters

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -913,18 +914,101 @@ IsFinalRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
+    }
+
+    auto type = inherited.getResolvedType(i, TypeResolutionStage::Structural);
+    auto kp = type->getKnownProtocol();
```
Is it really worth adding a new TypeBase::getKnownProtocol()? It appears to only used in this one place.

kavon replied:
I found 3 and then added another use to help clean up some code. I kind of like it. 🤷🏼

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -157,22 +157,73 @@ bool TypeBase::isMarkerExistential() {
+  if (isTypeParameter()) {
+    auto canType = getCanonicalType();
+    auto &ctx = canType->getASTContext();
+    IsNoncopyableRequest req {canType};
```
The generic signature is not part of the caching key so that might be why it's busted. However it's not worth caching GenericSignature::requiresProtocol(); it's extremely fast.

kavon replied:
Yeah removing the caching fixed this! I wasn't sure if the call into the requirement machine was fast enough to be used uncached.

kavon commented:
@swift-ci please smoke test

kavon committed:
[nfc] remove stray whitespace

kavon committed:
[Sema] bring back fast-path for isNoncopyable(DC)

Turns out my caching was busted and `requiresProtocol` is fast enough
to go without it.

kavon committed:
[Sema] simplify checks in isNoncopyable(DC)

Calling `mapTypeIntoContext` on n interface type with no type
parameters is a no-op. In addition, we can avoid calling
`getGenericEnvironmentOfContext` twice, which does a search of the
DeclContext to find it.

kavon committed:
[Sema] simplify `IsNoncopyableRequest::evaluate`

kavon committed:
[Sema] simplify `TypeResolver::resolveInverseType`

kavon commented:
@swift-ci please test

kavon committed:
[nfc] fix bad merge

kavon commented:
@swift-ci please test

kavon committed:
bump number of diags in `0161-issue-49119.swift`

Seems we already emit 6 duplicate diagnostics for this test, so what's
one more?

I'm not sure what change I made triggered this additional diagnostic.

kavon commented:
@swift-ci please test and merge
